### PR TITLE
Migrate diploid path finding

### PR DESCRIPTION
**Pull request by pjbradbury**
## Description

_Provide a summary of your changes including motivation and context.
If these changes fix a bug or resolves a feature request, be sure to link to that issue._

## Type of change

_What type of changes does your code introduce? Put an `x` in boxes that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated relevant documentation

## Changelog entry

_Please add a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note

```

### CODE REVIEW

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        .validate { require(File(it).isDirectory) {"$it is not a valid directory."} }
+
+    val probCorrect by option(help = "The probability that a mapped read was mapped correctly")
+        .double()
+        .default(0.99)
+        .validate { require(it in 0.5..1.0) {"prob-correct must be between 0.5 and 1.0"} }
+
+    val probSameGamete by option(help = "The probability of transitioning to the same gamete in the next reference range")
+        .double()
+        .default(0.99)
```
Comment (lynnjo) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt:69]: If user provides a number > 1.0, does it default to 1.0?  OR is this an error?

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        .validate { require(it in 0.0..1.0) {"inbreeding-coefficient must be between 0.0 and 1.0"} }
+
+    val maxReadsPerKb by option(help = "ReferenceRanges with more than max-reads-per-kb will not be imputed.")
+        .int()
+        .default(1000)
+        .validate { require(it > -1) {"max-reads-per-kb must be a positive integer."} }
+
+    val useLikelyAncestors by option(help="Use only the most likely ancestors of each sample for path finding.")
+        .boolean()
+        .default(false)
```
Comment (lynnjo) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt:94]: Should there be a check or warning in the code if useLikelyAncestors is true, but likelyAncestorFile is blank?  I see you do check if this file is blank in savePath().  And appendParentStats checks if the file exists (can that check for ""?)

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+        .int()
+        .default(Int.MAX_VALUE)
+
+    val minCoverage by option(help = "If use-likely-ancestors = true, use the fewest number of ancestors that together have this proportion of mappable reads.")
+        .double()
+        .default(1.0)
+        .validate { require(it in 0.5..1.0) {"min-coverage must be between 0.5 and 1.0"} }
+
+    val likelyAncestorFile by option(help="If useLikelyAncestors is true, a record of the ancestors used for each sample will be written to this file.")
+        .default("")
```
Comment (lynnjo) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt:106]: Should there be a check or warning in the code if useLikelyAncestors is true, but likelyAncestorFile is blank? I see you do check if this file is blank in savePath(). And appendParentStats checks if the file exists (can that check for ""?)

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+    val threads by option(help = "number of threads used to find paths.").int().default(3)
+
+
+    private val myLogger = LogManager.getLogger(DiploidPathFinding::class.java)
+
+    private fun buildHaplotypeGraph(): HaplotypeGraph {
+        val listOfHvcfFilenames = File(hvcfDir).listFiles().filter { it.name.endsWith(".h.vcf") }.map { it.path }
```
Comment (lynnjo) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt:114]: Does this need to also check for h.vcf.gz files?

**Review by lynnjo — APPROVED** [review]

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+                            val priorCount = mergedList[hapids] ?: 0
+                            mergedList[hapids] = priorCount + count
+                        }
+                    }
+                    mergedList
+                }
+            }
+        }
+
+        fun altHeaderMetadataToVCFHeaderLine(altHeaderData: AltHeaderMetaData): VCFAltHeaderLine {
```
Comment (tcasstevens) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt:312]: Can this be a utility that I can use also?

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/HaploidPathFinding.kt [hunk 6]
```diff
@@ -110,6 +110,7 @@ class HaploidPathFinding : CliktCommand(help = "Impute haploid paths") {
     }
 
     override fun run() {
+        //TODO determine how samples with no haplotype are being handled
```
Comment (tcasstevens) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/HaploidPathFinding.kt:113]: Are you planning to do this in this pull request?

File: src/main/kotlin/net/maizegenetics/phgv2/utils/VariantLoadingUtils.kt [hunk 7]
```diff
@@ -284,6 +285,38 @@ fun createHVCFRecord(assemblyTaxon: String, startPosition: Position, endPosition
     return vcb.make()
 }
 
+fun createDiploidHVCFRecord(sampleName: String, startPosition: Position, endPosition: Position, calls: List<String>, refAlleleStr: String): VariantContext {
```
Comment (tcasstevens) [inline src/main/kotlin/net/maizegenetics/phgv2/utils/VariantLoadingUtils.kt:288]: Do we plan to support higher ploids?

**Review by tcasstevens — APPROVED** [review]

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+ * 2. Build the HaplotypeGraph
+ * 3. For each entry in the keyfile (multithreaded)
+ *      a. Get the read mapping data
+ *      b. Use Viterbi algorithm to find the best path
+ *      c. Store the path (write hvcf)
+ *
+ * The samples in the keyfile are processed in parallel. Each thread (co-routine) processes a sample. Within sample
+ * processing is not multithreaded.
+ */
+class DiploidPathFinding: CliktCommand(help = "Impute best diploid path using read mappings.") {
```
[the remark this answers is not in the document]
Comment (pjbradbury) [reply src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt:44]: I think you are right. I added HaploidPathFinding as well.

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidEmissionProbability.kt [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+            for (haplotype in haplotypesInRefrange) {
+                val haplotypePair = UnorderedHaplotypePair(Pair(haplotype, null))
+                probabilityMap[haplotypePair] = haplotypePairProbability(haplotypePair, readSetCounts)
+            }
+            val haplotypePair = UnorderedHaplotypePair(Pair(null, null))
+            probabilityMap[haplotypePair] = haplotypePairProbability(haplotypePair, readSetCounts)
+        }
+
+        //take the natural log of the probabilities
+        rangeLnProbabilities = probabilityMap.mapValues { (_, pr) -> ln(pr) }
```
Comment (zrm22) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidEmissionProbability.kt:85]: Could this be done on the fly above when you are assigning the initial probabilities?  Would save a loop through the states.

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidEmissionProbability.kt [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    private fun createSampleToHaplotypeMap(): Map<SampleGamete, String> {
```
Comment (zrm22) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidEmissionProbability.kt:88]: Should this be a function of the Graph data structure?

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt (hunk 1 above)
Comment (zrm22) [reply src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt:69]: It should throw an error due to the require statement.  I believe this is tested in the unit tests.

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+    }
+
+    /**
+     * Takes a map of samples to file list as input. If the output directory contains an hvcf file for any sample,
+     * that sample will be removed from the map and the trimmed map will be returned.
+     */
+    private fun checkForExistingOutput(sampleToFiles: Map<String, List<String>>): Map<String, List<String>> {
+        val existingSampleFiles = File(outputDir).listFiles()
+        val existingSamples = existingSampleFiles.map {file ->
+            file.name.substringBefore(".h.vcf", "$$$")
```
Comment (zrm22) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt:160]: Should you just use an empty string here?  Or rather filter any name missing the .h.vcf before then just do a substringBefore without the second parameter?

**Review by zrm22 — APPROVED** [review]

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+        val myGraph = buildHaplotypeGraph()
+
+        //check onditions for useLikelyAncestor
+        if (useLikelyAncestors) {
+            if(maxAncestors >= myGraph.sampleGametesInGraph().size && minCoverage == 1.0)
+                myLogger.warn("UseLikelyAncestors is true but likelyAncestors will not be checked because minCoverage is 1.0 " +
+                        "and maxAncestors is >= the number of potential ancestors.")
+            else if (likelyAncestorFile.isBlank()) myLogger.warn("Likely ancestors will be determined for each sample " +
+                    "but information about the ancestors used will not be written to a file because no file name was provided.")
+        }
```
Comment (aberthel) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt:192]: Do we just want to warn here, or throw an error and exit? If we just warn then a user won't notice the message and will wonder why they didn't get the desired results.

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/PathFinderWithViterbiHMM.kt [hunk 13]
```diff
@@ -250,200 +254,156 @@ class PathFinderWithViterbiHMM(
...
-        /**
+    data class DiploidPath(val nodeList: List<DiploidPathNode>, val totalProbability: Double) {
+        fun terminatesIn(gametePair: Pair<SampleGamete, SampleGamete>) = nodeList.last().gametePair() == gametePair
+
+        fun terminalGametePair(): Pair<SampleGamete, SampleGamete> = nodeList.last().gametePair()
+    }
+
+
+    /**
+     * Creates a new path by appending gametePair to the previous path ana adding probability to totalProbability
```
Comment (aberthel) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/PathFinderWithViterbiHMM.kt:272]: Minor typo: "path ana adding" -> "path and adding"

**Review by aberthel — APPROVED** [review]

**Pull request merged**